### PR TITLE
FED-1721 Null safety preparation for class component default props / initial state

### PR DESCRIPTION
**Pull request by aaronlademann-wf**
## Motivation
Class based components use `defaultProps` and `initialState` to set the value of certain props / state fields when the class is constructed.

To map this pattern to null safety - we need props that are defaulted, and state that is initialized to be marked with a `/*late*/` hint to indicate that it will be set at the time of class construction, and a `/*!*/` hint to assist the null safety migrator so it knows that defaulted props / initialized state will never be null.

## Changes
1. Add two suggestor classes that share a lot of common functionality in order to add the aforementioned hints on prop field declarations that are defaulted, and on state field declarations that are initialized.
1. Add those suggestors to the existing `null_safety_migrator_companion` executable.
1. Write tests.

### What it does
**Before**
```dart
mixin FooProps on UiProps {
  String defaultedNullable;
  num defaultedNonNullable;
}
mixin FooState on UiState {
  String initializedNullable;
  bool initializedNonNullable;
}
class FooComponent extends UiStatefulComponent2<FooProps, FooState> {
  @override
  get defaultProps => (newProps()
    ..defaultedNullable = null
    ..defaultedNonNullable = 2.1
  );

  @override
  get initialState => (newState()
    ..initializedNullable = null
    ..initializedNonNullable = true
  );
  // ...
}
```

**After**
```dart
mixin FooProps on UiProps {
  /*late*/ String/*?*/ defaultedNullable;
  /*late*/ num/*!*/ defaultedNonNullable;
}
mixin FooState on UiState {
  /*late*/ String/*?*/ initializedNullable;
  /*late*/ bool/*!*/ initializedNonNullable;
}
class FooComponent extends UiStatefulComponent2<FooProps, FooState> {
  @override
  get defaultProps => (newProps()
    ..defaultedNullable = null
    ..defaultedNonNullable = 2.1
  );

  @override
  get initialState => (newState()
    ..initializedNullable = null
    ..initializedNonNullable = true
  );
  // ...
}
```

## Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

Please review: 

### QA Checklist
- [ ] Tests were updated and provide good coverage of the changeset and other affected code
- [ ] Manual testing was performed if needed
    - [ ] Steps from PR author: 
        
    - [ ] Anything falling under manual testing criteria [outlined in CONTRIBUTING.md][contributing-manual-testing]

## Merge Checklist
While we perform many automated checks before auto-merging, some manual checks are needed:
- [ ] A Frontend Frameworks Design member has reviewed these changes
- [ ] There are no unaddressed comments _- this check can be automated if reviewers use the "Request Changes" feature_
- [ ] _For release PRs -_ Version metadata in Rosie comment is correct

[contributing-review-types]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react_codemod/blob/master/CONTRIBUTING.md#manual-testing-criteria

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: test/dart3_suggestors/null_safety_prep/class_component_required_initial_state_test.dart [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+            
+              @override
+              render() => null;
+            }
+          '''),
+        );
+      });
+
+      test(
+          'patches defaulted props in legacy classes using component1 boilerplate',
```
Comment (greglittlefield-wf) [inline test/dart3_suggestors/null_safety_prep/class_component_required_initial_state_test.dart:211]: copypasta
```suggestion
          'patches initializes state in legacy classes using component1 boilerplate',
```

File: test/dart3_suggestors/null_safety_prep/class_component_required_initial_state_test.dart [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+                ..initializedNullable = null
+                ..initializedNonNullable = 2.1
+                ..anotherInitializedNonNullable = 1.1
+                ..initializedNonNullableFn = () {}
+                ..initializedNonNullableList = []
+              );
+            
+              @override
+              render() => null;
+            }
```
Comment (greglittlefield-wf) [inline test/dart3_suggestors/null_safety_prep/class_component_required_initial_state_test.dart:92]: Same comment as in the props test

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Made some comments, but they're mostly minor. This looks awesome!

File: lib/src/dart3_suggestors/null_safety_prep/class_component_required_default_props.dart [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+/// the type to assist the `nnbd_migration:migrate` script when it attempts to
+/// infer a prop's nullability.
+///
+/// **Optionally**, an [sdkVersion] can be passed to the constructor.
+/// When set to a version that opts-in to Dart's null safety feature,
+/// the `late` / `?` type modifiers will be actual modifiers rather
+/// than commented hints. This should only be done using an explicit opt-in
+/// flag from the executable as most consumers that have migrated to null-safety
+/// will have already run this script prior to the null safety migration and thus
+/// the `/*late*/` / `/*?*/` hints will already be converted to actual modifiers.
```
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/dart3_suggestors/null_safety_prep/class_component_required_default_props.dart:42]: Ticket for this: https://jira.atl.workiva.net/browse/FED-2554

**Review by greglittlefield-wf — APPROVED** [review]
+1, gonna take a break then do some QAing!

**Review by greglittlefield-wf — CHANGES_REQUESTED** [review]
Just a couple things found while QAing. Tested on a few repos and the output looks fantastic!

**Review by greglittlefield-wf — APPROVED** [review]
+10

@Workiva/release-management-p

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**